### PR TITLE
add action to kick off refreshing catalogs

### PR DESCRIPTION
**Pull request by kellmarie**
**Merge checklist:**
- [x] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [x] `make static` has been run to update webpack bundling if any static content was updated
- [x] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [x] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [x] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [Travis](https://travis-ci.org/github/edx/edx-enterprise), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

All catalogs are refreshing:
Screen Shot 2020-10-20 at 11 43 09 AM

Catalogs that failed to refresh:
Screen Shot 2020-10-20 at 11 44 12 AM

Mixture of successfully refreshing and failing to refresh catalogs:
Screen Shot 2020-10-20 at 11 40 31 AM

### CODE REVIEW

File: enterprise/api_client/enterprise_catalog.py [hunk 1]
```diff
@@ -169,6 +170,35 @@ def get_content_metadata(self, enterprise_customer, enterprise_catalogs=None):
 
         return list(content_metadata.values())
 
+    @JwtLmsApiClient.refresh_token
+    def refresh_catalogs(self, enterprise_catalogs):
```
Comment (bbaker6225) [inline enterprise/api_client/enterprise_catalog.py:174]: I don't think we should allow refreshing multiple catalogs in this way since it could unintentionally cause weird behavior if the catalogs share a catalog query. It's not the best UX, but I think it's reasonable to say that if you only want to update a few catalogs you just individually update them.
If we do want to support this however, I think we'd need to update the backend to take multiple catalogs and update the set of distinct catalog queries between them all

File: enterprise/admin/actions.py [hunk 2]
```diff
@@ -59,3 +63,34 @@ def export_as_csv(modeladmin, request, queryset):  # pylint: disable=unused-argu
...
+
+def refresh_catalog(self, request, queryset):  # pylint: disable=unused-argument
+    """
+    Kicks off background running tasks for refreshing catalogs
+    """
+    catalog_client = EnterpriseCatalogApiClient(user=request.user)
+    refreshed_catalogs, failed_to_refresh_catalogs = catalog_client.refresh_catalogs(queryset)
+
+    # display catalog and task ids that were successfully started refreshing
+    updated_message = ''
```
Comment (bbaker6225) [inline enterprise/admin/actions.py:76]: It probably simplifies a lot if we restrict the updates to a single catalog, but the string concatenation for updated_message is hard to follow and read for me. It's not perfectly DRY but might be cleaner to write out the admin message in larger chunks?

Comment (bbaker6225) [reply enterprise/admin/actions.py:76]: I also think it could be helpful to throw the task id into the success message if we can. Doesn't matter for ECS but saves a little digging for any developer who winds up doing this and then wants to check flower/splunk to follow the progress

File: enterprise/api_client/enterprise_catalog.py (hunk 1 above)
Comment (kellmarie) [reply enterprise/api_client/enterprise_catalog.py:174]: Well is there a way to get django actions to work so that what gets acted on isn't through a bulk selection? And the catalogs sharing a query isn't an issue for the daily jobs that run?

Comment (bbaker6225) [reply enterprise/api_client/enterprise_catalog.py:174]: Sure, just give an error message if `queryset.count() > 1`. And the daily job does the idea of distinct catalogs mentioned above: https://github.com/edx/enterprise-catalog/blob/4fe13a5b20912422b5463edc720bb78af32e9862/enterprise_catalog/apps/catalog/management/commands/update_content_metadata.py#L56

File: enterprise/admin/actions.py (hunk 2 above)
Comment (kellmarie) [reply enterprise/admin/actions.py:76]: Task id is in the success message. I posted screenshots of what the messages look like in the PR descriptions.

Comment (bbaker6225) [reply enterprise/admin/actions.py:76]: Ah nice, I missed it at the bottom. Ignore that comment then

File: enterprise/api_client/enterprise_catalog.py (hunk 1 above)
Comment (kellmarie) [reply enterprise/api_client/enterprise_catalog.py:174]: Actually how does refreshing the multiple catalogs individually by selecting them one by one address the shared catalogs weirdness issue? If I were to individually click on catalog A and refresh it and then click on catalog B and refresh it one after the other would the window of time it would take me to do the clicking be large enough that we won't have to worry about the weirdness happening like we do for making multiple api calls in a loop?

Comment (bbaker6225) [reply enterprise/api_client/enterprise_catalog.py:174]: It would still happen. I don't like either way so I don't think it _really_ matters which way we go. It just feels weirder to me to allow multiple at once when we don't really support that, especially when this seems like it would only ever be used for a single catalog at a time by ECS/developers

Comment (kellmarie) [reply enterprise/api_client/enterprise_catalog.py:174]: I think it would be weirder to have a UI that is designed to be used for bulk selecting but we would need to somehow communicate with people to only select one catalog at a time. But even if we do communicate that we don't really have a way to say that what they should actually be doing is waiting for the previously refreshed catalog task to finish refreshing before selecting another. Because telling them to refresh them individually wouldn't actually solve the issue about worrying out data weirdness happening.

Maybe we should just change the refresh endpoint to allow for multiple catalogs so we can do what the daily job does and only refresh the distinct queries? That way we can try to avoid any data weirdness and also not have to try communicating on the front end to select things individually and then wait for the task to finish even though they might not actually know where to look to know if the task had finished.

Comment (iloveagent57) [reply enterprise/api_client/enterprise_catalog.py:174]: My two cents:
* It's an internal tool, and my _guess_ is that it will mostly be used to update only a single catalog at a time.  We can communicate to the users of this tool that they're probably better off only updating one at a time.
* I think it's reasonable for this client (or any other) to make a series of POSTs to `refresh_metadata` for a series of catalogs.
* We should modify the enterprise-catalog service to more gracefully deal with this type of situation, somehow ensuring that we don't bother updating the same piece of content metadata multiple times within a short time window.

**Review by iloveagent57 — APPROVED** [review]
Let's give it a whirl.
